test(parser): track number of allocations

camchenry opened the pull request:
- related: https://github.com/oxc-project/backlog/issues/5

This PR adds a reliable method of tracking the number of allocations made while running the parser. This includes both arena allocations (via a new feature), as well as system allocations (by creating a new global allocator). Reductions in these numbers should correlate to real-world performance improvements that aren't quantified in CodSpeed currently. This will also help prevent regressions where we accidentally allocate lots more memory than we expect.

Originally I had included total system bytes allocated, but it was tricky to get this number to match exactly between platforms. I opted not try and make this perfect but instead use the total number of allocations which is a good proxy for bytes anyway.

camchenry commented:
* **#12555** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12555?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/12555?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=12555).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_allocator/src/allocator.rs:
```diff
@@ -571,6 +597,12 @@ impl Allocator {
+    /// Returns the total number of allocations that have been made in this [`Allocator`] instance.
+    #[cfg(feature = "track_allocations")]
+    pub fn num_alloc(&self) -> usize {
+        NUM_ALLOC.load(Relaxed)
```
I've never entirely got my head around atomic ordering, but I think `Relaxed` here may not be sufficient. You want to ensure all the increments have been committed before reading the total. For the same reason, I'm not sure if `Relaxed` where the increments occurs needs to be something stronger.

Or maybe sleeping for a short time before loading the total count would work, to allow everything to "catch up".

overlookmotel reviewed (commented):
This is great.

In addition to the comments below:

There are also many arena allocations that occur without going through the `Allocator` APIs. Notably, `Vec` performs allocations and reallocations directly via `Bump`, by going through the `Alloc` trait (`alloc.alloc()` and `alloc.grow()` calls).

https://github.com/oxc-project/oxc/blob/5f50bc3126b3353c4a284a8a8049165b267bed40/crates/oxc_allocator/src/vec2/raw_vec.rs#L93-L115

https://github.com/oxc-project/oxc/blob/5f50bc3126b3353c4a284a8a8049165b267bed40/crates/oxc_allocator/src/vec2/raw_vec.rs#L742-L766

These allocations / reallocations aren't getting counted, and they're a significant proportion of arena allocations.

So the ideal would be to increment `num_alloc` in `Alloc::alloc` + `Alloc::grow` as well. But problem is that in those methods you only have access to the `Bump`, not the `Allocator`.

A terrible hack would be:

```rs
impl Alloc for Bump {
    #[inline(always)]
    fn alloc(&self, layout: Layout) -> NonNull<u8> {
        #[cfg(feature = "track_allocations")]
        unsafe {
            use crate::Allocator;
            use std::{
                mem::offset_of,
                ptr,
                sync::atomic::{AtomicUsize, Ordering},
            };
            #[expect(clippy::cast_possible_wrap)]
            const OFFSET: isize = (offset_of!(Allocator, num_alloc) as isize)
                - (offset_of!(Allocator, bump) as isize);
            let num_alloc_ptr = ptr::from_ref(self).byte_offset(OFFSET).cast::<AtomicUsize>();
            let num_alloc = num_alloc_ptr.as_ref().unwrap_unchecked();
            num_alloc.fetch_add(1, Ordering::SeqCst);
        }

        self.alloc_layout(layout)
    }
}
```

This is completely unsound! It relies on `Bump` always being wrapped inside an `Allocator`, which can't be statically proven. But in practice we know that's always the case in Oxc - we never use `Bump` on its own. This unsoundness would be unacceptable in production code, but as it's behind a cargo feature which is only used in this internal tool, I think it'd be OK.

When I finally finish up replacing `Bump` with our own custom allocator ([WIP](https://github.com/oxc-project/oxc/tree/05-16-feat_allocator_replace_bumpalo_)) we could build allocation counting into it, and remove this hack.

---

Note: Counting the number and total byte size of arena *reallocations* separately would be a really good measure. Allocation in arena is very cheap, but reallocation is *way* more expensive (see comment below). Of the stats for arena allocations, that's one of the measures which should be a priority to reduce.

overlookmotel commented on tasks/allocs/src/lib.rs:
```diff
@@ -0,0 +1,131 @@
+// is the same across different platforms for the purposes of tracking allocations.
+#[expect(unsafe_code, clippy::undocumented_unsafe_blocks)]
+unsafe impl GlobalAlloc for TrackedAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let ret = unsafe { mimalloc_safe::MiMalloc.alloc(layout) };
+        if !ret.is_null() {
+            NUM_ALLOC.fetch_add(1, SeqCst);
+        }
+        ret
+    }
```
Is mimalloc workable on all platforms we expect people to develop Oxc on? I seem to remember seeing elsewhere in codebase that we gate its use on some platforms - though I'm not sure why.

overlookmotel replied:
> I did consider this, but I found that the number of system allocations increased when I defaulted to the system allocator and there was more variance in the number of allocations. When I switched to mimalloc, it stabilized the number of allocations across my local machine and CI, but still kept some variance in the number of bytes allocated. Maybe if we can track down all the sources of non-determinism we could default back to the system allocator.

Very interesting. Why would there be a different number of allocations depending on the allocator? The code requesting allocations is the same. Doesn't make sense to me! But there's clearly something I'm missing.

But how the hell would we figure out where the difference is coming from? `#[track_caller]` on allocation functions, log every call, and then diff the logs? I have no idea at all if that'd work.

And just to clarify: When you say "there was more variance in the number of allocations", do you mean there was variance between different runs *on the same machine*, or there was a difference between local and CI?

> Regarding platform support: mimalloc is well supported across macOS, Linux, and Windows as far as I can tell. There were some issues with executing illegal instructions on some platforms (like Raspberry Pi), but I think those are fixed now: #11045, #11743.

👍 Red herring. Sorry!

camchenry commented on tasks/allocs/src/lib.rs:
```diff
@@ -0,0 +1,131 @@
+
+    let options = ParseOptions { parse_regular_expression: true, ..ParseOptions::default() };
+
+    // Warm-up by parsing each file first, and then measuring the actual allocations.
+    // TODO: For some reason, this causes some variance in the total system allocations, specifically
+    // for checker.ts and antd.js. I tried to track this down but wasn't able to figure out why.
+    for file in files.files() {
+        Parser::new(&allocator, &file.source_text, file.source_type).with_options(options).parse();
+    }
```
Ah now that you mention this, I feel like this must be it. That matches exactly the behavior I was seeing. Maybe it would be best to just do a very large allocation up front? I think I'll leave it this way for now though, as it should ensure it works as expected even if we change the files later.

camchenry commented:
> So the ideal would be to increment `num_alloc` in `Alloc::alloc` + `Alloc::grow` as well. But problem is that in those methods you only have access to the `Bump`, not the `Allocator`.
> 
> A terrible hack would be: [...]
> This is completely unsound! It relies on `Bump` always being wrapped inside an `Allocator`, which can't be statically proven. But in practice we know that's always the case in Oxc - we never use `Bump` on its own. This unsoundness would be unacceptable in production code, but as it's behind a cargo feature which is only used in this internal tool, I think it'd be OK.

I think I can live with this, if you're okay with a little bit of extra feature-gated code in these methods. I think the effort will be worth it, as this will help us unlock a new avenue of performance optimization as well as preventing memory/performance regressions.

> When I finally finish up replacing `Bump` with our own custom allocator ([WIP](https://github.com/oxc-project/oxc/tree/05-16-feat_allocator_replace_bumpalo_)) we could build allocation counting into it, and remove this hack.

This sounds like a great path forward to me.

overlookmotel commented:
> I think I can live with this, if you're okay with a little bit of extra feature-gated code in these methods. I think the effort will be worth it, as this will help us unlock a new avenue of performance optimization as well as preventing memory/performance regressions.

I can live with monstrous unsoundness, if you can! 😄 I agree it's a metric we can likely leverage to make improvements on perf.

overlookmotel commented:
Oh, just one more thing. There's an annoyance with cargo features.

We run tests with `--all-features`, and we don't really want this code active in tests. If nothing else, it'll slow them down due to all the atomic ops in `Allocator`, and if we're adding the terrible unsound hack to `Alloc` trait, then we *definitely* don't want that code running in any circumstances other than `oxc_allocs`.

The crappy "solution" we've found for feature-gating certain code that's only used for oxlint JS plugins is to have 2 features called `oxlint2` and `disable_oxlint2`. All the code which we want compiled for `napi/oxlint2` is gated with:

```rs
#[cfg(all(feature = "oxlint2", not(feature = "disable_oxlint2")))]
fn gated() { /* ... */ }
```

So `gated` is not compiled with `--all-features`. You have to just enable `oxlint2` feature by itself.

This is crap. But it's the best we've come up with so far. So unfortunately I suggest using the same hack here in `oxc_allocator` crate - `track_allocations` and `disable_track_allocations` features.

To double-protect against hack code being compiled in production builds, I'd recommend adding `disable_track_allocations` to default features in `oxc_allocator` and then disabling default features for `oxc_allocator` in `oxc_allocs`.

camchenry commented:
> This is crap. But it's the best we've come up with so far. So unfortunately I suggest using the same hack here in `oxc_allocator` crate - `track_allocations` and `disable_track_allocations` features.
> 
> To double-protect against hack code being compiled in production builds, I'd recommend adding `disable_track_allocations` to default features in `oxc_allocator` and then disabling default features for `oxc_allocator` in `oxc_allocs`.

Done. However, I wasn't able to add `disable_track_allocations` as a default feature, as you can't override the default features in `oxc_allocs` then apparently:

```
error: failed to load manifest for workspace member `oxc/tasks/allocs`
Caused by:
  error inheriting `oxc_allocator` from workspace root manifest's `workspace.dependencies.oxc_allocator`

Caused by:
  `default-features = false` cannot override workspace's `default-features`
```

camchenry commented on tasks/allocs/src/lib.rs:
```diff
@@ -0,0 +1,131 @@
+    }
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        unsafe { mimalloc_safe::MiMalloc.realloc(ptr, layout, new_size) }
+    }
```
Okay I'm tracking reallocs separately now. It appears the grow-in-place behavior is very platform dependent (or at least allocator dependent). mimalloc seems to always return a different pointer? And I get different results for the macOS system allocator and system allocator in CI with the `ubuntu-latest` image.

Example:
<img width="946" height="341" alt="image" src="https://github.com/user-attachments/assets/01eadf95-e002-4158-ad81-9818aab5f493" />

Boshen commented on tasks/allocs/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "oxc_allocs"
```
`allocs` is too vague, maybe `oxc_track_memory_allocations` and `track_memory_allocations` for the directory name.

You can add a just command to shorten typing: `just allocs`.

camchenry replied:
Fair. I added the `just allocs` alias, but renamed the task to `oxc_track_memory_allocations`.

Boshen commented on tasks/allocs/allocs_parser.snap:
👍 really nice!

Boshen commented on tasks/allocs/src/lib.rs:
```diff
@@ -0,0 +1,153 @@
+use std::sync::atomic::{AtomicUsize, Ordering::SeqCst};
+
+#[global_allocator]
+static GLOBAL: TrackedAllocator = TrackedAllocator;
```
I wonder whether we can get https://docs.rs/dhat/latest/dhat working as well. Just a thought though, no changes required.

camchenry committed:
test(parser): track number of allocations (#12555)

- related: https://github.com/oxc-project/backlog/issues/5

This PR adds a reliable method of tracking the number of allocations made while running the parser. This includes both arena allocations (via a new feature), as well as system allocations (by creating a new global allocator). Reductions in these numbers should correlate to real-world performance improvements that aren't quantified in CodSpeed currently. This will also help prevent regressions where we accidentally allocate lots more memory than we expect.

Originally I had included total system bytes allocated, but it was tricky to get this number to match exactly between platforms. I opted not try and make this perfect but instead use the total number of allocations which is a good proxy for bytes anyway.
